src/Compilers/CSharp/Portable/Symbols/FunctionPointerTypeSymbol.cs

```diff
@@ -0,0 +1,264 @@
+                    diagnostics.Add(ErrorCode.ERR_InvalidFunctionPointerCallingConvention, syntax.CallingConvention.GetLocation(), syntax.CallingConvention.Text);
+                }
+
+                RefKind refKind = RefKind.None;
+                TypeWithAnnotations returnType;
+
+                if (syntax.Parameters.Count == 0)
+                {
+                    returnType = TypeWithAnnotations.Create(typeBinder.CreateErrorType());
+                }
+                else
+                {
+                    var returnTypeParameter = syntax.Parameters[^1];
+
+
+                    var modifiers = returnTypeParameter.Modifiers;
+                    if (modifiers.Count > 0)
+                    {
+                        for (int i = 0; i < returnTypeParameter.Modifiers.Count; i++)
+                        {
+                            var modifier = modifiers[i];
+                            switch (modifier.Kind())
+                            {
+                                case SyntaxKind.RefKeyword when refKind == RefKind.None:
+                                    if (modifiers.Count > i + 1 && modifiers[i + 1].Kind() == SyntaxKind.ReadOnlyKeyword)
+                                    {
+                                        i++;
+                                        refKind = RefKind.RefReadOnly;
+                                    }
+                                    else
+                                    {
+                                        refKind = RefKind.Ref;
+                                    }
+
+                                    break;
+
+                                case SyntaxKind.RefKeyword:
+                                    Debug.Assert(refKind != RefKind.None);
+                                    // A return type can only have one '{0}' modifier.
+                                    diagnostics.Add(ErrorCode.ERR_DupReturnTypeMod, modifier.GetLocation(), modifier.Text);
+                                    break;
+
+                                default:
+                                    // '{0}' is not a valid function pointer return type modifier. Valid modifiers are 'ref' and 'ref readonly'.
+                                    diagnostics.Add(ErrorCode.ERR_InvalidFuncPointerReturnTypeModifier, modifier.GetLocation(), modifier.Text);
+                                    break;
+                            }
+                        }
+                    }
+
+                    returnType = typeBinder.BindType(returnTypeParameter.Type, diagnostics, basesBeingResolved);
+
+                    if (returnType.IsVoidType() && refKind != RefKind.None)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_NoVoidHere, returnTypeParameter.GetLocation());
+                    }
+                }
+
+                var signature = new FunctionPointerMethodSymbol(parent, callingConvention, refKind, returnType);
```

Comment:
>callingConvention [](start = 72, length = 17)

Do we really need to reflect an invalid calling convention in the symbol? We reported an error, I think it would be fine to use CallingConvention.Default after that.

Reply:
> We reported an error, I think it would be fine to use CallingConvention.Default after that.

I considered that, but my thought was that this would eventually be exposed to consumers, so I wanted to explicitly indicate that we had a bad value.

---
In reply to: [359559897](https://github.com/dotnet/roslyn/pull/40431#discussion_r359559897) [](ancestors = 359559897)